### PR TITLE
Crossinghighways

### PR DESCRIPTION
**Pull request by Rub21**
Crossing highways is working fine:  relative to : https://github.com/osmlab/osmlint/issues/31#issuecomment-172001446

cc. @geohacker @batpad @maning

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -156,3 +156,21 @@ test('missingHighwaysUS', function(t) {
...
+  t.plan(1);
+  logInterceptor();
+  processors.crossingHighways(opts, mbtiles, function() {
+    var logs = logInterceptor.end();
+    for (var i = 0; i < logs.length; i++) {
+      var geoJSON = JSON.parse(logs[i]);
+      t.comment('Pass: ' + (i + 1));
+      if (geoJSON.features.length > 0) {
+        t.equal(geoJSON.features[0].properties._osmlint, 'crossinghighways', 'Should be crossinghighways');
+        //geoJSON.features[0].geometry.type = LineString, Point, MultiPOint
```
Comment (Rub21) [inline test/test.js]: @geohacker  when we have the more than tree types of output, how should we do the testing?  in this case the output of Crossinghighways validator is  `LineString, Point, MultiPoint`.   what do you reckon?

Comment (geohacker) [inline test/test.js]: @Rub21 we can check if the type is either a LineString, Point or MultiPoint?

File: validators/crossingHighways/map.js [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    'tertiary': true,
+    'trunk': true,
+    'residential': true,
+    'unclassified': true,
+    'service': true
+  };
+
+  layer.features.forEach(function(val) {
+    if (preserve[val.properties.highway] && (val.geometry.type === 'LineString' || val.geometry.type === 'MultiLineString') && val.properties.layer === undefined) {
+      val.properties._osmlint = 'crossinghighways';
```
Comment (geohacker) [inline validators/crossingHighways/map.js]: Can you move this line to right before writing the result to stdout?

File: validators/crossingHighways/map.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+  var output = {};
+  var result = [];
+
+  bboxes.forEach(function(bbox) {
+    var overlaps = traceTree.search(bbox);
+    overlaps.forEach(function(overlap) {
+      if (bbox[4] !== overlap[4]) {
+        var intersect = turf.intersect(highways[overlap[4]], highways[bbox[4]]);
+        if (intersect !== undefined && (intersect.geometry.type === 'Point' || intersect.geometry.type === 'MultiPoint')) {
+          var highwaycoors = _.flatten(highways[overlap[4]].geometry.coordinates);
```
Comment (geohacker) [inline validators/crossingHighways/map.js]: @Rub21 Can you rename `highwaycoors` to `highwayCoord` to keep them uniform like [here](https://github.com/osmlab/osmlint/search?utf8=%E2%9C%93&q=Coord).

File: validators/crossingHighways/map.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+          highwaycoors.concat(_.flatten(highways[bbox[4]].geometry.coordinates));
+          var intersectcoors = _.flatten(intersect.geometry.coordinates);
+          if (_.difference(highwaycoors, intersectcoors).length === highwaycoors.length) {
```
Comment (geohacker) [inline validators/crossingHighways/map.js]: @Rub21 Can you tell me more about this line? If the length difference between the highway coordinates and the intersection coordinates is same as length of highway coordinates, what does that mean?

File: validators/crossingHighways/map.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+            } else {
+              output[overlap[4].toString().concat(bbox[4])] = intersect;
+            }
+          }
+        }
+      }
+    });
+  });
+
+  _.each(output, function(v) {
```
Comment (geohacker) [inline validators/crossingHighways/map.js]: @Rub21 You should use http://underscorejs.org/#values for this one.

File: validators/crossingHighways/map.js (hunk 4 above)
Comment (Rub21) [inline validators/crossingHighways/map.js]: @geohacker : 
On these two lines
`var highwaycoors = _.flatten(highways[overlap[4]].geometry.coordinates);
 highwaycoors.concat(_.flatten(highways[bbox[4]].geometry.coordinates));` 

I am creating a `highwaycoors` it is an array of  number from all coordinates in both roads. Aso I am getting a array of  numbers.  from `intersect` node.
` var intersectcoors = _.flatten(intersect.geometry.coordinates);` 

and then,  check if the intersection node isn't in a node which in both of highways. if the `intersectcoors` exits in `highwaycoors` it should be a intersection in a point which exist in both highways. if is not  it means `intersectcoors` is a new intersect coordinates.

`_.difference` is just for compare two array, if `intersectcoors` does not exit  in `highwaycoors` `_.difference(highwaycoors, intersectcoors).length  `and ` highwaycoors.length` have the same length

**Comment by Rub21** [comment]
@geohacker  I have cleaned up according your comment and fix some others stuff  as well. 

I think now is good time merge.

**Pull request merged**